airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/BigQueryDestination.kt

```diff
@@ -0,0 +1,394 @@
+            )
+
+            val gcsDestination: BaseGcsDestination = object : BaseGcsDestination() {}
+            val gcsJsonNodeConfig = getGcsJsonNodeConfig(config)
+            return gcsDestination.check(gcsJsonNodeConfig)
+        } catch (e: Exception) {
+            val message = StringBuilder("Cannot access the GCS bucket.")
+            if (!missingPermissions.isEmpty()) {
+                message
+                    .append(" The following permissions are missing on the service account: ")
+                    .append(java.lang.String.join(", ", missingPermissions))
+                    .append(".")
+            }
+            message.append(
+                " Please make sure the service account can access the bucket path, and the HMAC keys are correct.",
+            )
+
+            log.error(e) { message.toString() }
+            throw ConfigErrorException(
+                "Could not access the GCS bucket with the provided configuration.\n",
+                e,
+            )
+        }
+    }
+
+    /**
+     * Returns a [AirbyteMessageConsumer] based on whether the uploading mode is STANDARD INSERTS or
+     * using STAGING
+     *
+     * @param config
+     * - integration-specific configuration object as json. e.g. { "username": "airbyte",
+     * "password": "super secure" }
+     * @param catalog
+     * - schema of the incoming messages.
+     */
+    override fun getConsumer(
+        config: JsonNode,
+        catalog: ConfiguredAirbyteCatalog,
+        outputRecordCollector: Consumer<AirbyteMessage>
+    ): AirbyteMessageConsumer? {
+        throw UnsupportedOperationException("Should use getSerializedMessageConsumer")
+    }
+
+    @Throws(Exception::class)
+    override fun getSerializedMessageConsumer(
+        config: JsonNode,
+        catalog: ConfiguredAirbyteCatalog,
+        outputRecordCollector: Consumer<AirbyteMessage>
+    ): SerializedAirbyteMessageConsumer {
+        val uploadingMethod = getLoadingMethod(config)
+        val defaultNamespace = getDatasetId(config)
+        setDefaultStreamNamespace(catalog, defaultNamespace)
+        val disableTypeDedupe = getDisableTypeDedupFlag(config)
+        val datasetLocation = getDatasetLocation(config)
+        val projectId = config[bqConstants.CONFIG_PROJECT_ID].asText()
+        val bigquery = getBigQuery(config)
+        val rawNamespaceOverride = getRawNamespaceOverride(bqConstants.RAW_DATA_DATASET)
+
+        addAllStringsInConfigForDeinterpolation(config)
+        val serviceAccountKey = config[bqConstants.CONFIG_CREDS]
+        if (serviceAccountKey != null) {
+            // If the service account key is a non-null string, we will try to
+            // deserialize it. Otherwise, we will let the Google library find it in
+            // the environment during the client initialization.
+            if (serviceAccountKey.isTextual) {
+                // There are cases where we fail to deserialize the service account key. In these
```

Comment:
why do we ever try to deserialize the string ourselves?

Reply:
This is the most ugly one where the serviceAccountKey was at one point text vs json. so trying to deduce which one it is.